realsense2_camera/src/realsense_node_factory.cpp

```diff
@@ -57,12 +74,110 @@ void RealSenseNodeFactory::getDevice(rs2::device_list list)
 		else
 		{
 			bool found = false;
+
+			if (_accel_orientation.empty()) // generate accel orientation vector from camera quaternion
+			{
+				try {
+					geometry_msgs::TransformStamped transformStamped = 
+									_tf_buffer.lookupTransform(_tf_link_name, "base_link", ros::Time(0));
+
+					tf2::Quaternion q_orig, q_rot;
+					// universal gravity vector
+					q_orig.setValue(0.0, -9.81, 0.0);
+					tf2::convert(transformStamped.transform.rotation , q_rot);
+					// Calculate the new orientation
+					q_orig = q_rot.inverse() * q_orig * q_rot; 
+
+					std::stringstream ss;
+					ss << q_orig.x() << " " << q_orig.y() << " " << q_orig.z();
+					_accel_orientation = ss.str();
+
+					ROS_DEBUG_STREAM("Computing from tf2 quaternion for " << _camera_name << " rotation: " << transformStamped.transform.rotation
+									<< ", accel: " << ss.str() << " " << q_orig.w());
+				} catch (tf2::TransformException e) {
+					ROS_ERROR_STREAM(e.what());
+					return;
+				}
+			}
+
+			// correctly decode and fill rs2_vector for expected median accelerometer orientation vector
+			rs2_vector accel_orientation_vec;
+			sscanf (_accel_orientation.c_str(), "%f%f%f", &accel_orientation_vec.x, &accel_orientation_vec.y, &accel_orientation_vec.z);
+			ROS_DEBUG_STREAM("Looking for " << _camera_name << " device with orientation " << accel_orientation_vec);
+
 			for (auto&& dev : list)
 			{
 				auto sn = dev.get_info(RS2_CAMERA_INFO_SERIAL_NUMBER);
 				ROS_DEBUG_STREAM("Device with serial number " << sn << " was found.");
-				if (_serial_no.empty() || sn == _serial_no)
+
+				if (!_serial_no.empty() && sn == _serial_no)
+				{
+					_device = dev;
+					_serial_no = sn;
+					found = true;
+					break;
+				} else if (_serial_no.empty() && !_accel_orientation.empty()) // auto determine correct realsense orientation
 				{
+					rs2::pipeline pipe;
+					rs2::config cfg;
+					cfg.enable_device(sn);
+					cfg.enable_stream(RS2_STREAM_ACCEL);
+
+					// attempt to create pipe stream for upto max attempts
+					const int MAX_ATTEMPTS = 5;
+					bool success = false;
+					int no_attempts = 0;
+					while (no_attempts < MAX_ATTEMPTS)
+					{
+						try
+						{
+							pipe.start(cfg);
+							success = true;
+							break;
+						} catch (...)
+						{
+							no_attempts ++;
+							ROS_DEBUG_STREAM("Pipe failed to create on Device serial no " << sn << "device busy. Waiting..." << no_attempts);
+							ros::Duration(2).sleep();
+							continue;
+						}
+					}
+
+					if (!success)
+					{
+						// all max attempts exhausted for the current device
```

Comment:
consider to publish a warning in this case as this is not expected to happen, right?

Reply:
Done. Will commit the code.